### PR TITLE
feat: refactor row format with slice formant and single/multi slice row format

### PR DESCRIPTION
**Pull request by tobegit3hub**
* Refactor slice row format
* Support single slice row format

### CODE REVIEW

File: hybridse/include/codec/fe_row_codec.h [hunk 1]
```diff
@@ -217,6 +219,80 @@ class RowFormat {
...
+
+class SingleSliceRowFormat : public RowFormat {
+ public:
+    // TODO: Add deconstructor
+    SingleSliceRowFormat(const Schema* schema) {
+        slice_format_ = new SliceFormat(schema);
+    }
+
+    SingleSliceRowFormat(std::vector<const Schema*> schemas) {
+        Schema merge_schema;
```
Comment (jingchen2222) [inline hybridse/include/codec/fe_row_codec.h]: temp variable merge_schema

File: hybridse/src/vm/schemas_context.cc [hunk 2]
```diff
@@ -510,19 +512,23 @@ const codec::Schema* SchemasContext::GetOutputSchema() const {
...
+    std::vector<const hybridse::codec::Schema*> schemas;
     for (const auto& source : schema_sources_) {
         if (source->GetSchema() == nullptr) {
             LOG(WARNING) << "Source schema is null";
             return;
         }
-        row_formats_.emplace_back(codec::RowFormat(source->GetSchema()));
+        schemas.push_back(source->GetSchema());
     }
+    row_format_ = new codec::MultiSlicesRowFormat(schemas);
```
Comment (jingchen2222) [inline hybridse/src/vm/schemas_context.cc]: if (row_format) {
  delete row_format;
}

File: hybridse/src/codegen/context.h [hunk 3]
```diff
@@ -172,7 +172,7 @@ class CodeGenContext {
 
     const vm::SchemasContext* schemas_context_;
     const codec::Schema* parameter_types_;
-    const codec::RowFormat parameter_row_format_;
+    codec::RowFormat* parameter_row_format_ = nullptr;
```
Comment (jingchen2222) [inline hybridse/src/codegen/context.h:176]: if (parameter_row_format) {
delete parameter_row_format;
}

File: hybridse/include/codec/fe_row_codec.h [hunk 4]
```diff
@@ -217,6 +217,89 @@ class RowFormat {
     uint32_t str_field_start_offset_;
 };
 
+class RowFormat {
+ public:
```
Comment (dl239) [inline hybridse/include/codec/fe_row_codec.h:221]: add virtual deconstructor

File: hybridse/include/codec/fe_row_codec.h [hunk 5]
```diff
@@ -217,6 +217,89 @@ class RowFormat {
...
+    virtual bool GetStringColumnInfo(size_t schema_idx, size_t idx, StringColInfo* res) const = 0;
+    virtual const ColInfo* GetColumnInfo(size_t schema_idx, size_t idx) const = 0;
+    virtual size_t GetSliceId(size_t schema_idx) const = 0;
+};
+
+class MultiSlicesRowFormat : public RowFormat {
```
Comment (dl239) [inline hybridse/include/codec/fe_row_codec.h:227]: what's the different between `SliceFormat` and `RowFormat`. And is has some confused that the relationship between `MultiSlicesRowFormat` and `SliceFormat`

File: hybridse/src/vm/schemas_context.cc [hunk 6]
```diff
@@ -141,7 +143,11 @@ void SchemasContext::Clear() {
         delete ptr;
     }
     schema_sources_.clear();
-    row_formats_.clear();
+    if (row_format_) {
+        delete row_format_;
+        row_format_ = nullptr;
+    }
+    delete row_format_;
```
Comment (dl239) [inline hybridse/src/vm/schemas_context.cc]: repeated delete

File: hybridse/include/codec/fe_row_codec.h [hunk 7]
```diff
@@ -217,6 +217,89 @@ class RowFormat {
...
+    }
+
+    ~SingleSliceRowFormat() {
+        offsets_.clear();
+        if (slice_format_) {
+            delete slice_format_;
+        }
+    }
+
+    explicit SingleSliceRowFormat(std::vector<const Schema*> schemas) {
```
Comment (dl239) [inline hybridse/include/codec/fe_row_codec.h]: pass const reference vec

File: hybridse/src/codec/fe_row_codec_test.cc [hunk 8]
```diff
@@ -405,7 +405,7 @@ TEST_F(CodecTest, ManyCol) {
     }
 }
 
-TEST_F(CodecTest, RowFormatTest) {
+TEST_F(CodecTest, MultiSlicesRowFormatTest) {
```
Comment (tobegit3hub) [inline hybridse/src/codec/fe_row_codec_test.cc]: Change to test `SliceFormat`.

File: hybridse/src/codegen/aggregate_ir_builder.cc [hunk 9]
```diff
@@ -677,15 +677,16 @@ base::Status AggregateIRBuilder::BuildMulti(const std::string& base_funcname,
     // compute current row's slices
     for (auto& pair : agg_col_infos_) {
         size_t schema_idx = pair.second.schema_idx;
-        auto iter = used_slices.find(schema_idx);
+        size_t slice_idx = schema_context_->GetRowFormat()->GetSliceId(schema_idx);
```
Comment (tobegit3hub) [inline hybridse/src/codegen/aggregate_ir_builder.cc]: For some UT, GetRowFormat null be nullptr.

**Review by dl239 — APPROVED** [review]

File: hybridse/src/vm/physical_op.cc [hunk 10]
```diff
@@ -281,7 +281,10 @@ Status PhysicalConstProjectNode::InitSchema(PhysicalPlanContext* ctx) {
     SchemaSource* project_source = schemas_ctx_.AddSource();
 
     SchemasContext empty_schemas_ctx;
-    return InitProjectSchemaSource(project_, &empty_schemas_ctx, ctx, project_source);
+    CHECK_STATUS(InitProjectSchemaSource(project_, &empty_schemas_ctx, ctx, project_source));
+    CHECK_STATUS(ctx->InitFnDef(project_, &schemas_ctx_, true, &project_),
```
Comment (aceforeverd) [inline hybridse/src/vm/physical_op.cc:285]: comment for later check #1445

**Review by aceforeverd — APPROVED** [review]

**Pull request merged**